### DISCUSSION TITLE
Request: Add ImpHash

### DISCUSSION THREAD

**Discussion by neu5ron** [zeek/zeek#321] [Ideas]
I am not sure the level of effort to add Import Address Table(IAT) Hashing (ImpHash) to Zeek, similar to how there is MD5, SHA1, SHA256, etc..
This would be a great feature to add to the pe.log and or files.log(for PEs). I think it would be effective and very useful for a few reasons (and probably more that others could think of):
1) Additional malware detection via "pivoting" off of the ImpHash. Both for internal to external traffic but also lateral movement (east/west).
ImpHash is useful (to a decent degree) for clustering. For example different samples may have different md5/sha1/sha256 hashes, but if its a similar malware family then its possible they are using the same ImpHash (due to the malware's functionality). I have seen even some recent APTs hitting the news of big breaches but have the same ImpHash in each (despite completely different targets/campaigns).
2) Legitimate Microsoft PE's being renamed and transferred (over the wire) during lateral movement. This scenario happens a lot when an actor moves a legitimate Microsoft PE, such as Sysinternals PsExec, but changes the name of the file so as not to appear as "psexec.exe".
One would say to just use the md5/sha1/sha256 of the file, however with different versioning and "modifications" that would make it a non factor to use. Also, its rare that the IAT changes in legitimate binaries through a few version changes.
This is especially useful for smb.log!

Thanks for your time!

**Comment by neu5ron** [comment]
Especially since the pe.log already obtains ["has_import_table" ](https://github.com/zeek/zeek/blob/956674745bbfe4b8e8a29323f06ccffa9c18b3b9/scripts/base/files/pe/main.bro#L114) could possibly make this a if statement somewhere saying  if "has_import_table" and a redef for enable imphash -- then perform imphashing and add to log.

**Comment by 0xekez** [comment]
As far as I understand, looking at an implementation [here](https://github.com/erocarrera/pefile/blob/bbf28920a71248ed5c656c81e119779c131d9bd4/pefile.py#L4217) this would just be a matter of adding a .bif function that could compute the ImpHash of some table of values. Like you mentioned, there is already a function to compute the MD5 hash, so it would just be a matter of doing a little parsing of the [import table](https://docs.zeek.org/en/stable/scripts/base/files/pe/consts.bro.html#id-PE::directories) to get them in the right form. You could then add an optional string value to the log that holds the ImpHash.

The only thing I'm not sure about are rather or not we can safely modify log formats without breaking other people's code. Do you have any thoughts on this @jsiwek? It's also a little unclear to me what the syntax would be to retreive the import table once you had established that it exists, but I'm sure I could work that out.

If that all sounds good I'm happy to work on this.

**Comment by jsiwek** [comment]
@ZekeMedley historically, we've been ok with extending log files with *new* fields, but hesitate to change the semantics or type of an existing field.  Sounds like this work entails the former, so not a big deal.

**Comment by grigorescu** [comment]
I looked at this when writing the PE analyzer. Today, the import address table is NOT parsed by the PE analyzer. I did have a version that could do this; src/file_analysis/analyzer/pe/pe-file-idata.pac was left from this version, but it only worked with PE files from a few compilers because it didn't correctly account for the import address table (IAT) being out of order.

There's a lot more data available in PE files, but due to the structure of the file format, this would need a fair amount of overhauling of the analyzer. Really, it just needs to be rewritten in Spicy.

I had been working on writing up a proof of concept in Python, which could then be ported to Binpac. See: https://github.com/grigorescu/pynpac/blob/master/examples/parse_pe.py

**Reply by jsiwek** [reply]
> There's a lot more data available in PE files, but due to the structure of the file format, this would need a fair amount of overhauling of the analyzer. Really, it just needs to be rewritten in Spicy.

Coincidentally, I just started writing a Spicy parser for PE: https://github.com/zeek/spicy/compare/topic/jsiwek/pe.  Seems to parse import/export tables fine for a few files I tested against.  Let me know if you notice anything important I overlooked about the structure or compiler-specific differences/weirdness that should be considered.

I've been using https://github.com/erocarrera/pefile for checking/referencing output of the Spicy PE parser as I worked on it, but not so much referencing the internal logic -- couldn't tell how much some complexities were conservative/guesswork to deal w/ corner-case PE weirdness, so for now, mostly did the simplest thing that seemed to work in the Spicy parser as far as I understood the file format specification

**Reply by neu5ron** [reply]
awwesome work!

yara pe module may be good reference too https://github.com/VirusTotal/yara/blob/master/libyara/modules/pe/pe.c

**Reply by benjeems** [reply]
Myself and @davisshannon from Splunk just built a package to add imphash to pe.log. We did this on top of the Spicy PE analyzer https://github.com/zeek/spicy-analyzers/tree/main/analyzer/pe . We'll release the package shortly after a little more testing for edge cases, but it seems to work well so far. Anyone wishing to help test - please ping one of us.
